### ISSUE TITLE
Failed config

### ISSUE THREAD

**Issue by MariuszWoszczynski** [Wumfie/ESPHome-HPMA115S0-Sensor-Component#3]
Hi.
I downloaded 4 files, put them in custom_components/hpma115s0_esphome/ directory, reset HA and this integration in ESPHome is not working. What could be the cause?

![obraz](https://user-images.githubusercontent.com/115429759/195141669-fa6c42c5-a324-4920-930d-f2e0a420b174.png)
![obraz](https://user-images.githubusercontent.com/115429759/195142371-c896586d-0b71-48af-9387-e05afbaa906d.png)

**Comment by tom1422** [comment]
What version of ESPHome are you running? (I think there have been some changes in the new versions)

**Comment by MariuszWoszczynski** [comment]
2022.9.4

**Comment by tom1422** [comment]
Ok I think I have got it working again (I was running a very old version of esphome). I will update the code in the main repo and have just tested it on my dev board with no issues. 
There might be issues with the string handling which I haven't fixed but I think it depends on your development board so let me know about any problems. 
Thanks, Tom.

**Comment by MariuszWoszczynski** [comment]
I uploaded the new files and the same thing....

**Comment by MariuszWoszczynski** [comment]
All yaml:
![obraz](https://user-images.githubusercontent.com/115429759/195177133-08e7a3ab-1728-4593-8466-b7f680ce18bc.png)

**Comment by tom1422** [comment]
Try upload/compile, whats the error?

**Comment by MariuszWoszczynski** [comment]
the same
![obraz](https://user-images.githubusercontent.com/115429759/195179475-a9c8cc77-a463-4d9a-ab0c-85fe8d3115b8.png)

**Comment by MariuszWoszczynski** [comment]
shit......
I created folder hpma115s0_esphome in custom_component but in main HA folder, not ESPHome 🤦‍♂️

**Comment by tom1422** [comment]
Ok I have just pushed another update which should fix the string handling issues.

**Comment by tom1422** [comment]
> shit...... I created folder hpma115s0_esphome in custom_component but in main HA folder, not ESPHome 🤦‍♂️

😆I guess that explains it then. Anyway I have pushed some needed changes (there were actually some real issues which have been addressed, so thank you for notifying me). Let me know if it works now (with the code I pushed ~2 mins ago)

**Comment by MariuszWoszczynski** [comment]
[logs_wezel1_compile.txt](https://github.com/tom1422/ESPHome-HPMA115S0-Sensor-Component/files/9759303/logs_wezel1_compile.txt)

**Comment by MariuszWoszczynski** [comment]
there are some errors

**Comment by tom1422** [comment]
I see in the compile log there is an error which should be fixed now (there is no longer a c_str() function). Could you try with the latest? (Not from the releases tab - download master as zip instead)

**Comment by MariuszWoszczynski** [comment]
this is the last one
[logs_wezel1_compile.txt](https://github.com/tom1422/ESPHome-HPMA115S0-Sensor-Component/files/9759337/logs_wezel1_compile.txt)

**Comment by tom1422** [comment]
Sorry I missed that, I have just pushed another update, let me know how it goes

**Comment by MariuszWoszczynski** [comment]
Thanks, you are great:)
It compiled and programmed my nodemcu, but....

It does not see the sensor. Plugged into GPIO 9 and 10.
[logs_wezel1_logs.txt](https://github.com/tom1422/ESPHome-HPMA115S0-Sensor-Component/files/9759461/logs_wezel1_logs.txt)

**Comment by tom1422** [comment]
Looks like a timeout (1-Success, 2-Timeout, 3-NACK, 4-Malformed) - check the pinout of the sensor (make sure rx and tx are the right way around - it wont harm the device getting it backwards). And check input voltage and stuff.

**Comment by MariuszWoszczynski** [comment]
Everything looks ok:/

**Comment by tom1422** [comment]
TX should go to RX on the HPMA115S0 (and RX on ESP to TX on HPMA). Check the datasheet if you are unsure. [2313714.pdf](https://www.farnell.com/datasheets/2313714.pdf) . Also check this: [https://esphome.io/components/uart.html](https://esphome.io/components/uart.html), specifically the Hardware UARTs section. I would try pins 13 and 15.

It runs perfectly fine on my sensor. The only other thing would be to increase the waitTime from 300ms to something higher in the hpma115s0_esphome.h file (but this is unlikely to change anything).

**Comment by MariuszWoszczynski** [comment]
This is obvious:)
I checked on other pins 1 and 3, 13 and 15, 9 and 10 and it did not work on any configuration. I checked the sensor at work today with a UART/USB converter and the sensor is sending data.

**Comment by tom1422** [comment]
Would you mind enabling the debug flag on the uart section in your ESPhome config file. I have made another folder called test with some extra code which should help with debugging (it waits 8 seconds before trying to connect so it lets you see debug info). Would you mind posting logs for it? (Also ignore the initial error with 0 AND 0).

**Comment by MariuszWoszczynski** [comment]
This is a log from install and first run
[logs_wezel1_upload.txt](https://github.com/tom1422/ESPHome-HPMA115S0-Sensor-Component/files/9765593/logs_wezel1_upload.txt)

**Comment by MariuszWoszczynski** [comment]
and this is log from run
[logs_wezel1_logs.txt](https://github.com/tom1422/ESPHome-HPMA115S0-Sensor-Component/files/9765632/logs_wezel1_logs.txt)

after power reset, it`s working........

**Comment by tom1422** [comment]
Is the sensor still working?
It could be that the sensor has a long startup time so it isn't detected initially. In the logs I don't see any uart debug output so could you try it again with the debug flag on? [https://esphome.io/components/uart.html#debugging](https://esphome.io/components/uart.html#debugging)

**Comment by MariuszWoszczynski** [comment]
Yes, it still works. It looks like the nodemcu requires a reset after programming.
I added simple debug:

uart:
  id: uart_bus
  tx_pin: 15 #Change tx_pin and rx_pin to the ones you're using
  rx_pin: 13
  baud_rate: 9600
  debug:

Here is a log:
[logs_wezel1_upload.txt](https://github.com/tom1422/ESPHome-HPMA115S0-Sensor-Component/files/9766629/logs_wezel1_upload.txt)
Off course cot working without reset

**Comment by MariuszWoszczynski** [comment]
Afrer power reset on nodemcu:
[logs_wezel1_logs.txt](https://github.com/tom1422/ESPHome-HPMA115S0-Sensor-Component/files/9766635/logs_wezel1_logs.txt)

**Comment by MariuszWoszczynski** [comment]
Yesterday files working too after reset nodemcu.

I wonder why this is happening. When I install the station externally and want to change the soft, it will have to be manually reset, which will be a bother. Maybe it is possible to force a software reset of the nodemcu if there is no connection to the sensor? I don't know if this will do anything though, because the power on the sensor will remain on reset (I'm powering it directly from 5V USB).

**Comment by tom1422** [comment]
Mine is mounted outside with its own PSU and i update via OTA so it resets without turning off the dust sensor.

**Comment by tom1422** [comment]
I might give it a go with just the dust sensor and an ESP8266 (nodemcu) and see if it exhibits similar behavior. It might be a power supply issue or something. Esp-home supports wireless firmware upgrade which should reset it anyway if that's what I understand from your issue?

**Comment by tom1422** [comment]
Ok I just tried it with an external power supply and no issues (esp8266) and switching on the PSU at same time as ESP boot.

**Comment by MariuszWoszczynski** [comment]
I think here is a similar, unsolved problem:
https://arduino.stackexchange.com/questions/49240/honeywell-hpma115s0-sensor-not-start-again-after-stop/
